src/plugins/fs/f2fs.c

```diff
@@ -0,0 +1,395 @@
+#define SHRINK_DEPS_RESIZEF2FS_MASK (1 << SHRINK_DEPS_RESIZEF2FS)
+
+#define SHRINK_DEPS_LAST 1
+
+static const UtilDep shrink_deps[SHRINK_DEPS_LAST] = {
+    {"resize.f2fs", "1.12.0", "-V", "resize.f2fs\\s+([\\d\\.]+).+"}
+};
+
+static guint32 fs_mode_util[BD_FS_MODE_LAST+1] = {
+    /*   mkfs          wipe     check               repair        set-label     query           resize */
+    DEPS_MKFSF2FS_MASK, 0, DEPS_CHECKF2FS_MASK, DEPS_FSCKF2FS_MASK,   0,  DEPS_DUMPF2FS_MASK,  DEPS_RESIZEF2FS_MASK
+};
+
+#define UNUSED __attribute__((unused))
+
+/* option to get version was added in 1.11.0 so we need to cover situation
+   where the version is too old to check the version */
+static gboolean can_check_f2fs_version (UtilDep dep, GError **error) {
+    gboolean available = FALSE;
+    GError *loc_error = NULL;
+
+    available = bd_utils_check_util_version (dep.name, dep.version,
+                                             dep.ver_arg, dep.ver_regexp,
+                                             &loc_error);
+    if (!available) {
+        if (g_error_matches (loc_error, BD_UTILS_EXEC_ERROR, BD_UTILS_EXEC_ERROR_UTIL_UNKNOWN_VER)) {
+            /* assuming version of f2fs is too low to check version of f2fs */
+            g_clear_error (&loc_error);
+            g_set_error (error, BD_UTILS_EXEC_ERROR, BD_UTILS_EXEC_ERROR_UTIL_LOW_VER,
+                         "Too low version of %s. At least %s required.",
+                         dep.name, dep.version);
+            return FALSE;
+        }
+    } else {
+        /* just ignore other errors (e.g. version was detected but is still
+           too low) -- check_deps call below will cover this and create a
+           better error message for these cases */
+        g_clear_error (&loc_error);
+    }
+
+    return TRUE;
+}
+
+/**
+ * bd_fs_f2fs_is_tech_avail:
+ * @tech: the queried tech
+ * @mode: a bit mask of queried modes of operation (#BDFSTechMode) for @tech
+ * @error: (out): place to store error (details about why the @tech-@mode combination is not available)
+ *
+ * Returns: whether the @tech-@mode combination is available -- supported by the
+ *          plugin implementation and having all the runtime dependencies available
+ */
+gboolean bd_fs_f2fs_is_tech_avail (BDFSTech tech UNUSED, guint64 mode, GError **error) {
+    guint32 required = 0;
+    guint i = 0;
+
+    if (mode & BD_FS_TECH_MODE_SET_LABEL) {
+        g_set_error (error, BD_FS_ERROR, BD_FS_ERROR_TECH_UNAVAIL,
+                     "F2FS doesn't support setting label for an existing device.");
+        return FALSE;
+    }
+
+    if (mode & BD_FS_TECH_MODE_CHECK) {
+        if (!can_check_f2fs_version (deps[DEPS_CHECKF2FS], error))
+            return FALSE;
+    }
+
+    for (i = 0; i <= BD_FS_MODE_LAST; i++)
+        if (mode & (1 << i))
+            required |= fs_mode_util[i];
+
+    return check_deps (&avail_deps, required, deps, DEPS_LAST, &deps_check_lock, error);
+}
+
+/**BD_FS_ERROR_INVAL
```

Comment:
What's this `BD_FS_ERROR_INVAL` doing here? A copy pasta?

Reply:
Fixed.